File: ED del Pedro/ED/Haskell/Julio 2022/haskell/DiGraphDftTimer.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
-- Identity number (DNI if Spanish/passport if Erasmus):
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DiGraphDftTimer(
    diGraphDftTimer
) where

import           DataStructures.Dictionary.AVLDictionary as D
import           DataStructures.Graph.DiGraph
import           DataStructures.Set.BSTSet               as S

-- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
-- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
-- EXERCISE 3

diGraphDftTimer :: (Ord v) => DiGraph v -> (Dictionary v Int, Dictionary v Int)
diGraphDftTimer diGraph = aux vs diGraph (S.empty) (D.empty) (D.empty) 0
    where
        vs = vertices diGraph
        aux [] g visited arrivalT departureT _ = (arrivalT,departureT)
        aux (x:xs) g visited arrivalT departureT timer 
            | S.isElem x visited = aux xs g visited arrivalT departureT timer
            | otherwise = aux xs g visited' arrivalT' (D.insert x timer' departureT') (timer' + 1)
                where
                    (visited',arrivalT',departureT',timer') = checkSuccessors x g (S.insert x visited) (D.insert x timer arrivalT) departureT (timer+1) 

checkSuccessors :: (Ord v) => v -> DiGraph v -> Set v -> Dictionary v Int -> Dictionary v Int -> Int -> (Set v,Dictionary v Int, Dictionary v Int, Int)
checkSuccessors x g visited aT dT time = aux x sucs g visited aT dT time
    where
        sucs = successors g x
        aux v [] g s aT dT time = (s,aT,dT',time)
            where
                dT' = D.insert v time dT
        aux v (x:xs) g s aT dT time
            | S.isElem x s = aux v xs g s aT dT time
            | otherwise = aux v xs g s' aT' dT' (time'+1)
                where
                    (s',aT',dT',time') = aux x (successors g x) g (S.insert x s) (D.insert x time aT) dT (time+1)

File: ED del Pedro/ED/Java/Julio 2022/src/exercises/DiGraphDftTimer.java
/* ------------------------------------------------------------------------------
   -- Student's name:
   -- Student's group:
   -- Identity number (DNI if Spanish/passport if Erasmus):
   --
   -- Data Structures. Grado en Informática. UMA.
   -------------------------------------------------------------------------------
*/

package exercises;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.graph.DiGraph;
import dataStructures.set.HashSet;
import dataStructures.set.Set;

public class DiGraphDftTimer<V> {
  
  private int time;
  private Dictionary<V, Integer> arrivalD, departureD;
  private Set<V> visited;

// -- ESCRIBE TU SOLUCIÓN DEBAJO ----------------------------------------------
// -- WRITE YOUR SOLUTION BELOW  ----------------------------------------------
// -- EXERCISE 3

  private int checkSuccessors(V v, DiGraph<V> graph){
    // Miramos sus sucesores
    Set<V> sucesores = graph.successors(v);
    if(!sucesores.isEmpty()){
      // Como los sucesores no están vacíos, itero
      for(V suc: sucesores){
        if(!visited.isElem(suc)){
          visited.insert(suc);
          arrivalD.insert(suc,time);
          ++time;
          time = checkSuccessors(suc,graph);
          time++;
        }
      }
    }
    departureD.insert(v,time);
    return time;
  }

  public DiGraphDftTimer(DiGraph<V> diGraphg) {
    // TODO
    // Primero debo pasar por cada vértice y debo guardar por cada uno el tiempo de llegada y tiempo de salida
    time = 0;
    visited = new HashSet<>();
    arrivalD = new HashDictionary<>();
    departureD = new HashDictionary<>();
    // Ahora tendría que iterar sobre el grafo
    for(V vertex: diGraphg.vertices()){
      // Miro si esta visitado o no, porque si esta visitado no hago nada
      if(!visited.isElem(vertex)){
        // Si no ha sido visitado lo miramos
        visited.insert(vertex);
        // Actualizo tiempo de llegada
        arrivalD.insert(vertex,time);
        ++time;
        time = checkSuccessors(vertex,diGraphg);
        departureD.insert(vertex,time);
        time++;
      }
    }
  }

  public int arrivalTime(V v) {
    return arrivalD.valueOf(v);
  }

  public int departureTime(V v) {
    return departureD.valueOf(v);
  }
}
